src/flow/persistence/mod.rs

```diff
@@ -66,13 +75,100 @@ impl Parameters {
     /// `queue_capacity` indicates the number of packets that should be buffered until
     /// flushing, and `flush_timeout` indicates the length of time to wait before flushing
     /// anyway.
-    pub fn new(mode: DurabilityMode, queue_capacity: usize, flush_timeout: time::Duration) -> Self {
+    pub fn new(
+        mode: DurabilityMode,
+        queue_capacity: usize,
+        flush_timeout: time::Duration,
+        log_prefix: Option<String>,
+    ) -> Self {
         Self {
             queue_capacity,
             flush_timeout,
             mode,
+            log_prefix: log_prefix.unwrap_or(String::from("soup")),
         }
     }
+
+    /// The path that would be used for the given domain/shard pair's logs.
+    pub fn log_path(
+        &self,
+        node: &LocalNodeIndex,
+        domain_index: domain::Index,
+        domain_shard: usize,
+    ) -> PathBuf {
+        let filename = format!(
+            "{}-log-{}_{}-{}.json",
+            self.log_prefix,
+            domain_index.index(),
+            domain_shard,
+            node.id()
+        );
+
+        PathBuf::from(&filename)
+    }
+}
+
+/// Retrieves a vector of packets from the persistent log.
+pub fn retrieve_recovery_packets(
+    nodes: &DomainNodes,
+    domain_index: domain::Index,
+    domain_shard: usize,
+    params: &Parameters,
+    checktable: Rc<checktable::CheckTableClient>,
+) -> Vec<Box<Packet>> {
+    let mut packets = vec![];
+    for (_index, node) in nodes.iter() {
+        let node = node.borrow();
+        let local_addr = node.local_addr();
+        let global_addr = node.global_addr();
+        let path = params.log_path(&local_addr, domain_index, domain_shard);
+        let file = match File::open(&path) {
+            Ok(f) => f,
+            Err(ref e) if e.kind() == ErrorKind::NotFound => continue,
+            Err(e) => panic!("Could not open log file {:?}: {}", path, e),
+        };
+
+        BufReader::new(file)
+            .lines()
+            .flat_map(|line| {
+                let line = line.unwrap();
+                let entries: Vec<Records> = serde_json::from_str(&line).unwrap();
```

Comment:
If the system was part way through writing a line when it crashed, then deserialization may fail during recovery. It is worth handling this case.

Reply:
Done!